### PR TITLE
[geonode] a generic way to consume geonode api. fixes #31376

### LINKED ISSUE #31376: GeoNode provider does not work with newer versions of GeoNode
**Describe the bug**
Some valid GeoNode connections return no layers. I suspect it is because the QGIS provider needs to be updated for newer GeoNode versions. 

**How to Reproduce**

1. Create a GeoNode connection to http://geonode.inasafe.org/ (GeoNode 2.9)
1. See how it connects and expands to show available layers
1. Create a GeoNode connection to http://geonode.ingc.gov.mz/
1. Test the connection and note that it is a successful connection
1. Note that the connection does not expand to show any layers. 

**QGIS and OS versions**

QGIS 3
Mac OSX

### PR DESCRIPTION
**Pull request by myarjunar**
## Description

fix #31376 
  -  Remove the geonode version checking. Because I just found out that the geonode version has nothing to do with the format of the returned data.
  -  Redo the logic to construct OWS endpoint of layers.

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [ ] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

### CODE REVIEW

**Comment by nirvn** [comment]
Can we try to push this into 3.10.0? As it is, we'd ship QGIS with a semi broken geonode provider.

File: src/core/geocms/geonode/qgsgeonoderequest.cpp [hunk 1]
```diff
@@ -269,111 +269,106 @@ QList<QgsGeoNodeRequest::ServiceLayerDetail> QgsGeoNodeRequest::parseLayers( con
...
+          const QJsonDocument resourceUriDocument = QJsonDocument::fromJson( this->lastResponse() );
+          const QJsonObject resourceUriObject = resourceUriDocument.object();
+          const QVariantMap resourceUriMap = resourceUriObject.toVariantMap();
+          QVariantList resourceUriLinks = resourceUriMap.value( QStringLiteral( "links" ) ).toList();
+          QgsGeoNodeRequest::ServiceLayerDetail tempLayerStruct;
+          tempLayerStruct = parseOWSUrl( tempLayerStruct, resourceUriLinks );
+
+          // Avoid iterating all the layers to get the service url. Instead, generate a string format once we found one service url
+          // for every service (wms, wfs, xyz). And then use the string format for the other layers since they are identical.
+          if ( tempLayerStruct.server == "qgis-server" )
```
Comment (elpaso) [inline src/core/geocms/geonode/qgsgeonoderequest.cpp]:
```suggestion
          if ( tempLayerStruct.server == QStringLiteral( "qgis-server" ) )
```

File: src/core/geocms/geonode/qgsgeonoderequest.cpp [hunk 2]
```diff
@@ -269,111 +269,106 @@ QList<QgsGeoNodeRequest::ServiceLayerDetail> QgsGeoNodeRequest::parseLayers( con
...
+          {
+            wmsURLFormat = !tempLayerStruct.wmsURL.isEmpty() ? tempLayerStruct.wmsURL.replace( layerStruct.name, "%1" ) : "";
```
Comment (elpaso) [inline src/core/geocms/geonode/qgsgeonoderequest.cpp]:
```suggestion
            wmsURLFormat = ! tempLayerStruct.wmsURL.isEmpty() ? tempLayerStruct.wmsURL.replace( layerStruct.name, QStringLiteral( "%1" )  ) : QString();
```

Comment (elpaso) [reply src/core/geocms/geonode/qgsgeonoderequest.cpp]: following also.

File: src/core/geocms/geonode/qgsgeonoderequest.h [hunk 3]
```diff
@@ -213,6 +217,11 @@ class CORE_EXPORT QgsGeoNodeRequest : public QObject
      */
     void setProtocol( const QString &protocol );
 
+    /**
+     * Returns the updated ServiceLayerDetail struct with WMS/WFS/XYZ url.
+     */
+    QgsGeoNodeRequest::ServiceLayerDetail parseOWSUrl( QgsGeoNodeRequest::ServiceLayerDetail &layerStruct, const QVariantList &layerLinks );
```
Comment (elpaso) [inline src/core/geocms/geonode/qgsgeonoderequest.h]: I think the QGIS convention is to capitalize acronyms
```suggestion
    QgsGeoNodeRequest::ServiceLayerDetail parseOwsUrl( QgsGeoNodeRequest::ServiceLayerDetail &layerStruct, const QVariantList &layerLinks );
```

File: src/core/geocms/geonode/qgsgeonoderequest.h [hunk 4]
```diff
@@ -86,6 +88,8 @@ class CORE_EXPORT QgsGeoNodeRequest : public QObject
       QString wfsURL;
       //! XYZ tileserver URL for layer
       QString xyzURL;
+      //! Backend server (geoserver or qgis-server)
+      QString server;
```
Comment (m-kuhn) [inline src/core/geocms/geonode/qgsgeonoderequest.h]: Should this be an enum instead?

File: src/core/geocms/geonode/qgsgeonoderequest.cpp (hunk 1 above)
Comment (myarjunar) [reply src/core/geocms/geonode/qgsgeonoderequest.cpp]: ah yes of course, I must have missed some strings. thanks

File: src/core/geocms/geonode/qgsgeonoderequest.h (hunk 4 above)
Comment (myarjunar) [reply src/core/geocms/geonode/qgsgeonoderequest.h]: yeah I agree I think that is a better type. thanks!

**Comment by timlinux** [comment]
Anything more to do here?

File: src/core/geocms/geonode/qgsgeonoderequest.cpp [hunk 5]
```diff
@@ -269,111 +269,108 @@ QList<QgsGeoNodeRequest::ServiceLayerDetail> QgsGeoNodeRequest::parseLayers( con
...
-    else
+    QgsGeoNodeRequest::ServiceLayerDetail layerStruct;
+    const QVariantMap layerMap = layer.toMap();
+    QVariantList layerLinks = layerMap.value( QStringLiteral( "links" ) ).toList();
+
+    if ( layerMap.value( QStringLiteral( "typename" ) ).toString().isEmpty() )
     {
-      return layers;
+      const QStringList splitURL = layerMap.value( QStringLiteral( "detail_url" ) ).toString().split( '/' );
+      layerStruct.typeName = splitURL.at( splitURL.length() - 1 );
```
Comment (elpaso) [inline src/core/geocms/geonode/qgsgeonoderequest.cpp]: Please add a check that  "splitURL.length() - 1" is a valid index or we might have a crash.

**Review by elpaso — APPROVED** [review]
please just add the check for at() index validity and IMHO we are good to go

**Pull request merged**